uqcsbot/utils/snailrace_utils.py

```diff
@@ -0,0 +1,146 @@
+            self.step_number += 1
+
+    def __str__(self):
+        # Calculate the index of the snail emoji in the track
+        index = min(
+            int((self.position / SNAILRACE_TRACK_LENGTH) * SNAILRACE_TRACK_LENGTH_CHARS), 
+            SNAILRACE_TRACK_LENGTH_CHARS - 1
+        )
+
+        # Create the track string
+        track = SNAILRACE_TRACK_LENGTH_CHARS * [" "]
+        track[:index] = ["."] * index
+        track[index] = SNAILRACE_SNAIL_EMOJI
+        track_str = "".join(track)
+
+        return f"{str(self.racer_id).rjust(3)} |{track_str}| {self.member.display_name}"
+
+
+class SnailRaceState:
+    def __init__(self):
+        self.race_start_time = datetime.datetime.now()
+        self.racing = False
+        
+        self.racers = []
+        self.open_interaction = None
+
+    def is_racing(self) -> bool:
+
+        # Calculate max race time with a 5 second bias
+        max_race_len = SNAILRACE_OPEN_TIME + (SNAILRACE_STEP_TIME * SNAILRACE_TRACK_LENGTH) + 5
+        if (datetime.datetime.now() - self.race_start_time).total_seconds() >= max_race_len:
+            self.close_race()
+
+        return self.racing
+
+    def open_race(self, open_interaction: discord.Interaction):
+        # Start the entry
+        self.racing = True
+        self.open_interaction = open_interaction
+
+    def close_race(self):
+        # Close the race for a new one to start
+        self.racing = False
+        self.racers = []
+        self.open_interaction = None
+
+    def add_racer(self, racer: discord.Member) -> bool:
+        # Filter Unique Racers
+        for r in self.racers:
+            if r.member.id == racer.id:
+                return False
+
+        self.racers.append(SnailRacer(racer, len(self.racers) + 1))
+        return True
+    
+    async def race_start(self):
+        await self._start_racing(self.open_interaction)
+
+    async def _start_racing(self, interaction: discord.Interaction):
+        """
+        Start the race loop, this will be triggered after the entry has closed.
+        """
+
+        if not len(self.racers) > 0:
+            await interaction.channel.send(SNAILRACE_NO_START)
+            self.close_race()
+            return
+
+        # Write the first message to the channel which will be edited later
+        race_msg = await interaction.channel.send("BANG!")
+
+        # Loop until all racers have finished
+        while not all(r.position >= SNAILRACE_TRACK_LENGTH for r in self.racers):
+            for r in self.racers:
+                r.step()
+           
+            # Build the board and edit the race message with the new board
+            board = str(SNAILRACE_BOARD % "\n".join(str(r) for r in self.racers))
+            race_msg = await race_msg.edit(content=board)
+
+            # Wait a second before the next step
+            await asyncio.sleep(SNAILRACE_STEP_TIME)
+        
+        # Find who won
+        min_steps = SNAILRACE_MIN_STEP *SNAILRACE_TRACK_LENGTH
```

Comment:
General comment:

You may also wish to tiebreak by "distance on final round" - i.e. in the event that multiple people have the same number of steps, on their last round, whoever goes the furthest "past the finish line" should win.

This depends on how common ties are in common use, which is somewhat hard to test. May be worth returning to implement this after the command goes live for a few days and we can see what a race with 10+ people looks like.

Reply:
Yeah we'll have to see when its being put in use. It will be interesting to see how multiple people would react in a tie based outcome.

Will revisit this and update according to user reactions. UX testing in Prod is good.